### ISSUE TITLE
Magento 2.3.4 Quote Address Extension Attribute Issue

### ISSUE THREAD

**Issue by NathMorgan** [magento/magento2#26682]
### Preconditions (*)

1. Using Magento 2.3.4
2. USing a custom extension attribute on `Magento\Quote\Api\Data\AddressInterface` Example module here: https://github.com/NathMorgan/module-test-module

### Steps to reproduce (*)
1. Sending a post request to `estimate-shipping-methods` containing a `extension_attributes`

For example a post request to `/rest/default/V1/guest-carts/{cartId}/estimate-shipping-methods` with the following example data `{"address":{"street":["Example High Street"],"city":"Example City","region_id":null,"region":null,"country_id":"GB","postcode":"TES 01","company":"Example Company","telephone":"0000 00000","fax":null,"extension_attributes":{"test_attribute":"2"}}}`

### Expected result (*)
1. Above action completed successfully

### Actual result (*)
1. 500 Error with the message `Uncaught Error: Call to a member function setDiscounts() on array in /var/www/html/vendor/magento/module-sales-rule/Model/Quote/Discount.php: 117`

This issue seems to be introduced with this commit: https://github.com/magento/magento2/commit/6900c38941850706d9bff3aae76a428afae1343c

This seems to effect any extension attribute set on `Magento\Quote\Api\Data\AddressInterface` that is not a object

**Comment by NathMorgan** [comment]
Currently working on a fix

**Comment by Swahjak** [comment]
This seems to be specifically caused by https://github.com/magento/magento2/commit/6900c38941850706d9bff3aae76a428afae1343c#diff-0dba08a1fa5d66da0b3c252c4949907eR106 which populates the extension attribute. Then when it passes through https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Quote/Model/ShippingMethodManagement.php#L312 it will be converted to array by the `\Magento\Framework\Reflection\DataObjectProcessor`. Then the `addData` will blindly overwrite `extension_attributes` with an array value.

This could be a 'custom' implementation error only, because this will only happen if the discount collector is called before (any public function that depends on `getShippingMethods`) ShippingMethodManager.

On a side note; this would also cause errors if any other extension attribute is added that doesn't default to `null`.

**Comment by Swahjak** [comment]
Dug in a little deeper and in my opinion the solution / problem is this line https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Quote/Model/ShippingMethodManagement.php#L312 

I think there are two solutions to the problem:
1. Since 'extraAddressData' uses the `\Magento\Framework\Reflection\DataObjectProcessor` it would make sense to restore the data using the `Magento\Framework\Webapi\ServiceInputProcessor` to restore the data. This already has build in support for extension attributes. Although it would be more consistent if the `\Magento\Framework\Reflection\DataObjectProcessor` would also be replaced with `Magento\Framework\Webapi\ServiceOutputProcessor` (although the more or less give the same result).
2. Add a line right after the add / extract data line `$shippingAddress->setExtensionAttributes($address->getExtensionAttributes());`.
(3. Don't break stuff in minor releases)

The second option is the easiest to implement but doen't really 'fix' the issue, it just works around the issue in a non-standard way.

**Comment by NathMorgan** [comment]
@Swahjak Sadly the 2nd method is not as simple due to the fact that the `$address` param can be multiple diferent class types (That contain there own type of extension attributes). I've currently pushing up a change that was done to "fix" this (Only sets data if both the $address and the $shippingAddress extension attributes match)

**Comment by amenk** [comment]
@NathMorgan are you still working on this? Is there any WIP branch or workaround?

**Comment by amenk** [comment]
For our case the following patch seems to work. We just avoid overwriting the extension_attributes. Is is really necessary to overwrite them using extractAddressData() ? I think it is just a sideeffect and extension attributes should not be overwritten in that way


	--- ShippingMethodManagement.orig.php   2020-03-18 15:37:24.262000000 +0100
	+++ ShippingMethodManagement.php        2020-03-18 15:37:30.458462262 +0100
	@@ -297,7 +297,12 @@
	     {
		 $output = [];
		 $shippingAddress = $quote->getShippingAddress();
	-        $shippingAddress->addData($this->extractAddressData($address));
	+
	+        $addressData = $this->extractAddressData($address);
	+        if (array_key_exists('extension_attributes', $addressData)) {
	+            unset($addressData['extension_attributes']);
	+        }
	+        $shippingAddress->addData($addressData);
		 $shippingAddress->setCollectShippingRates(true);
	 
		 $this->totalsCollector->collectAddressTotals($quote, $shippingAddress);

**Comment by azambon** [comment]
I tried @amenk 's patch and it seems to work. My only doubt is whether the same thing should be done for "custom attributes" as well.

In case anyone else wants to try that same patch without editing the core, I turned it onto a diff that can be applied automatically with Composer Patches. Just save this diff into a file and configure Composer Patches to apply it to the magento/module-quote module:

```
diff --git a/Model/ShippingMethodManagement.php b/Model/ShippingMethodManagement.php
index 1b86f06..56edacf 100644
--- a/Model/ShippingMethodManagement.php
+++ b/Model/ShippingMethodManagement.php
@@ -297,7 +297,16 @@ class ShippingMethodManagement implements
     {
         $output = [];
         $shippingAddress = $quote->getShippingAddress();
-        $shippingAddress->addData($this->extractAddressData($address));
+
+        //Patch for Magento issue #26682
+        /** @see https://github.com/magento/magento2/issues/26682 */
+        $addressData = $this->extractAddressData($address);
+        if (array_key_exists('extension_attributes', $addressData)) {
+            unset($addressData['extension_attributes']);
+        }
+        $shippingAddress->addData($addressData);
+        //End patch
+
         $shippingAddress->setCollectShippingRates(true);

         $this->totalsCollector->collectAddressTotals($quote, $shippingAddress);
```

**Comment by amenk** [comment]
Thanks @azambon . You can also get a diff by adding .diff to the PRs URL:

https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/27338.diff

Can even be referenced from composer patches directly.

**Comment by azambon** [comment]
Hmm, are you sure that it works using the straight PR diff?
The issue that I see is the file path. In the PR diff the path is app/code/Magento/Quote/Model/ShippingMethodManagement.php, which is correct if one wants to apply the patch to a clone of the main Magento repo.
But if one wants to patch a regular Magento 2 installation where each module is a separate dependency ( https://devdocs.magento.com/guides/v2.3/install-gde/composer.html ), then the file path in the diff must be changed accordingly to refer to the root of the single component.

Am I missing something?

**Comment by amenk** [comment]
In composer-patches there is a level-parameter which was just invented for that :)

**Comment by azambon** [comment]
@amenk Didn't know about that parameter. I'll look into it. Thank you :+1:

**Comment by hkvega01** [comment]
I've tried to create a module to override the core model but didn't work, 

```
namespace Custom\FixShippingMethod\Model;

    class ShippingMethodManagement extends \Magento\Quote\Model\ShippingMethodManagement
    {
        /**
         * Get list of available shipping methods
         *
         * @param \Magento\Quote\Model\Quote $quote
         * @param \Magento\Framework\Api\ExtensibleDataInterface $address
         * @return \Magento\Quote\Api\Data\ShippingMethodInterface[]
         */
        public function getShippingMethods(Quote $quote, $address)
        {

            $writer = new \Zend\Log\Writer\Stream(BP . '/var/log/shippingmethodmanagement.log');
            $logger = new \Zend\Log\Logger();
            $logger->addWriter($writer);
            $logger->info('test test test');

            $output = [];
            $shippingAddress = $quote->getShippingAddress();
            //Patch for Magento issue #26682
            /** @see https://github.com/magento/magento2/issues/26682 */
            $addressData = $this->extractAddressData($address);
            if (array_key_exists('extension_attributes', $addressData)) {
                unset($addressData['extension_attributes']);
            }
            $shippingAddress->addData($addressData);
            //End patch
            $shippingAddress->setCollectShippingRates(true);

            $this->totalsCollector->collectAddressTotals($quote, $shippingAddress);
            $shippingRates = $shippingAddress->getGroupedAllShippingRates();
            foreach ($shippingRates as $carrierRates) {
                foreach ($carrierRates as $rate) {
                    $output[] = $this->converter->modelToDataObject($rate, $quote->getQuoteCurrencyCode());
                }
            }
            return $output;
        }
```

**Comment by ihor-sviziev** [comment]
Hi,

This issue was already fixed in 2.4-develop branch in https://github.com/magento/magento2/commit/fcd1b5808625364e09dc6aea2d9eb8c0ee209d25#diff-e2b9d1a5f4b9f3503a2f8dd072bc2d8aR360.

Additional test coverage will be also added in https://github.com/magento/magento2/pull/27338.

**Comment by magento-engcom-team** [comment]
Hi @NathMorgan. Thank you for your report. 
The issue has been fixed in magento/magento2#27338 by @amenk in 2.4-develop branch
Related commit(s):
 - [ea70c0ba65fd76c037c33ed77e1e1a40abbdb45a](https://github.com/magento/magento2/commit/ea70c0ba65fd76c037c33ed77e1e1a40abbdb45a)
 - [dcb59d31677bee2a9116bdab4ec17e1c1301b65a](https://github.com/magento/magento2/commit/dcb59d31677bee2a9116bdab4ec17e1c1301b65a)
 - [bc18a8fd5141a2748b4938e2468413ba7a3625ba](https://github.com/magento/magento2/commit/bc18a8fd5141a2748b4938e2468413ba7a3625ba)
 - [013bbc97b533856ac49607ed8a94c160dad32959](https://github.com/magento/magento2/commit/013bbc97b533856ac49607ed8a94c160dad32959)
 - [7d8638892f3e0b12c11d724dd2d6da8a1e27eaa2](https://github.com/magento/magento2/commit/7d8638892f3e0b12c11d724dd2d6da8a1e27eaa2)
 - [517bc6e0af4b69c3ab0a44dba914a4ca38ae40d8](https://github.com/magento/magento2/commit/517bc6e0af4b69c3ab0a44dba914a4ca38ae40d8)
 - [c29e3234b787eb3ea513ef49507c135b8bb921a0](https://github.com/magento/magento2/commit/c29e3234b787eb3ea513ef49507c135b8bb921a0)
 - [b60504e8dfe75dee6069adb5e12027e0b0f4fb3d](https://github.com/magento/magento2/commit/b60504e8dfe75dee6069adb5e12027e0b0f4fb3d)
 - [65c71534af4bffd86c09d7886770269470644509](https://github.com/magento/magento2/commit/65c71534af4bffd86c09d7886770269470644509)
 - [07538606525755713ffbe6691bd998735f640491](https://github.com/magento/magento2/commit/07538606525755713ffbe6691bd998735f640491)
 - [f133bff647e48d6f426cd732314c746ce7c80a70](https://github.com/magento/magento2/commit/f133bff647e48d6f426cd732314c746ce7c80a70)
 - [c38d99842781ec36accfcd40a3fefcfb4e76c9f9](https://github.com/magento/magento2/commit/c38d99842781ec36accfcd40a3fefcfb4e76c9f9)
 - [579712bf2a787206ecac4ef48a2a3c3c3277fc60](https://github.com/magento/magento2/commit/579712bf2a787206ecac4ef48a2a3c3c3277fc60)
 - [0518666727b48ce5b5a0373506312c53b74e452b](https://github.com/magento/magento2/commit/0518666727b48ce5b5a0373506312c53b74e452b)
 - [2aae8ae6a85ff9c06cfa7d86aa81ea0c81332e96](https://github.com/magento/magento2/commit/2aae8ae6a85ff9c06cfa7d86aa81ea0c81332e96)

The fix will be available with the upcoming 2.4.1 release.
